R/particle_filter_state.R

```diff
@@ -0,0 +1,220 @@
+##' @title Particle filter state
+##'
+##' @description Particle filter internal state. This object is not
+##'   ordinarily constructed directly by users, but via the
+##'   `$run_begin` method to [mcstate::particle_filter]. It provides
+##'   an advanced interface to the particle filter that allows
+##'   partially running the particle filter over part of the time
+##'   trajectory.
+##'
+##' This state object has a number of public fields that you can read
+##'   but must not write (they are not read-only so you *could* write
+##'   them, but don't).
+particle_filter_state <- R6::R6Class(
+  "particle_filter_state",
+  cloneable = FALSE,
+
+  private = list(
+    generator = NULL,
+    pars = NULL,
+    data = NULL,
+    data_split = NULL,
+    steps = NULL,
+    n_particles = NULL,
+    n_threads = NULL,
+    initial = NULL,
+    index = NULL,
+    compare = NULL,
+    save_restart_step = NULL,
+    save_restart = NULL,
+    current_step_index = 0L
+  ),
+
+  public = list(
+    ##' @field model The dust model generator being simulated (cannot be
+    ##'   re-bound)
+    model = NULL,
+
+    ##' @field complete Logical, indicating if the particle filter has
+    ##'   reached the end of the data.
+    complete = FALSE,
+
+    ##' @field history The particle history, if created with
+    ##'   `save_history = TRUE`. This is an internal format subject to
+    ##    change.
+    history = NULL,
+
+    ##' @field restart_state Full model state at a series of points in
+    ##'   time, if the model was created with non-`NULL` `save_restart`.
+    ##'   This is a 3d array as described in [mcstate::particle_filter]
+    restart_state = NULL,
+
+    ##' @field log_likelihood The log-likelihood so far. This starts at
+    ##'   0 when initialised and accumulates value for each step taken.
+    log_likelihood = NULL,
```

Comment:
To fix the SMC^2 issue we need a non-accumulating ll, which I think makes sense in this state object. Perhaps `current_log_likelihood`. This will store the value from the current application of compare only.

Does it make sense to add this now so I can use it when I try and apply the fix on the other PR? Or add that all at once?

Reply:
I can do that here, I think; that field is actually redundant atm

Reply:
Done in 54fc4e9